docker_aux/Rate4Site/3.0/share/man/man1/rate4site.1
.\" Automatically generated by Pod::Man 4.09 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.if !\nF .nr F 0
.if \nF>0 \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    if !\nF==2 \{\
.        nr % 0
.        nr F 2
.    \}
.\}
.\" ========================================================================
.\"
.IX Title "RATE4SITE 1"
.TH RATE4SITE 1 "2020-09-02" "3.0.0" "User Commands"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
rate4site \- detector of conserved amino\-acid sites
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
rate4site [\s-1OPTIONS\s0] \-s <\s-1MSA FILE\s0>
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The rate of evolution is not constant among amino acid sites: some positions evolve slowly and are commonly referred to as \*(L"conserved\*(R", while others evolve rapidly and are referred to as \*(L"variable\*(R". The rate variations correspond to different levels of purifying selection acting on these sites. The purifying selection can be the result of geometrical constraints on the folding of the protein into its 3D structure, constraints at amino acid sites involved in enzymatic activity or in ligand binding or, alternatively, at amino acid sites that take part in protein-protein interactions. Rate4Site calculates the relative evolutionary rate at each site using a probabilistic-based evolutionary model. This allows taking into account the stochastic process underlying sequence evolution within protein families and the phylogenetic tree of the proteins in the family. The conservation score at a site corresponds to the site's evolutionary rate.
.SH "METHODOLOGY"
.IX Header "METHODOLOGY"
The sole obligatory input to Rate4Site is an \s-1MSA\s0 file. The program then computes a phylogenetic tree that is consistent with the available \s-1MSA\s0 (the user can also input a pre-calculated tree). It then calculates the relative conservation score for each site in the \s-1MSA.\s0 This is carried out using either an empirical Bayesian method or a maximum likelihood method (Pupko et al., 2002). The differences between the two methods are explained in details in Mayrose et al (2004).
.SH "REFERENCES"
.IX Header "REFERENCES"
.IP "Mayrose, I., Graur, D., Ben-Tal, N., and Pupko, T. 2004. Comparison of site-specific rate-inference methods: Bayesian methods are superior. Mol Biol Evol 21: 1781\-1791." 4
.IX Item "Mayrose, I., Graur, D., Ben-Tal, N., and Pupko, T. 2004. Comparison of site-specific rate-inference methods: Bayesian methods are superior. Mol Biol Evol 21: 1781-1791."
.SH "OPTIONS"
.IX Header "OPTIONS"
.PD 0
.IP "\fB\-s \s-1MSA_FILE\s0\fR" 4
.IX Item "-s MSA_FILE"
.PD
The input sequence file name. The following formats are supported: Mase, Molphy, Phylip, Clustal, Fasta
.IP "\-t" 4
.IX Item "-t"
The input tree file name (in Newick format)
.IP "\fB\-o \s-1OUTPUT_FILE\s0\fR" 4
.IX Item "-o OUTPUT_FILE"
The results output file
.IP "\-a" 4
.IX Item "-a"
Reference sequence name in the \s-1MSA.\s0 The conservation scores are printed based on the amino-acids in this sequence.
.IP "\-k" 4
.IX Item "-k"
The number of discrete Gamma categories
.IP "\-m" 4
.IX Item "-m"
Evolutionary model. The following amino-acids models are supported:
.Sp
.Vb 1
\&        DAY (\-md), JTT (\-mj), REV (\-mr), aaJC (\-ma), LG (\-Ml), WAG (\-Mw) .
\&
\&        For nucleotides, the following models are supported: JC (\-mn), HKY (\-Mh), Tamura92 (\-Mt), GTR (\-Mg).
.Ve
.IP "\-b" 4
.IX Item "-b"
Branch lengths optimization flag:
.Sp
.Vb 1
\&        \-bn = no Branch lengths optimization
\&        
\&        \-bh = optimization using a homogenous model (no among\-site\-rate\-variation)
\&        
\&        \-bg = optimization using a Gamma model
.Ve
.IP "\-i" 4
.IX Item "-i"
Rate inference method flag:
.Sp
.Vb 1
\&        \-Im = rates are inferred using the maximum likelihood method
\&
\&        \-Ib = rates are inferred using the empirical Bayes method
.Ve
.IP "\-z" 4
.IX Item "-z"
Tree constructing method
.Sp
.Vb 1
\&        zj = Neighbor\-joining tree with Jukes\-Cantor distances
\&        
\&        zn = Neighbor\-joining tree with maximum likelihood distances
.Ve
.IP "\-h" 4
.IX Item "-h"
Short help message
.SH "AUTHOR"
.IX Header "AUTHOR"
Nir Ben-Tal <NirB@tauex.tau.ac.il>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.IP "Main website" 4
.IX Item "Main website"
<http://www.tau.ac.il/~itaymay/cp/rate4site.html>


src/programs/rate4site.py
import typing as t
from dataclasses import dataclass
import os
import re
from io import StringIO
import pandas as pd
from .program import Program

from dotenv import load_dotenv, find_dotenv

load_dotenv(find_dotenv())


@dataclass
class Rate4Site(Program):
    def __init__(self):
        self.name = "rate4site"
        self.program_exe = os.environ["rate4site"]
        self.cluster_program_exe = os.environ["cluster_rate4site"]
        self.input_param_name = "-s"
        self.output_param_name = "-o"
        self.module_to_load = "Rate4Site/Rate4Site-3.0"

    @staticmethod
    def parse_rates(rates_content: str) -> pd.DataFrame:
        """
        :param rates_content: a string given from rate4site output file
        :return: a parsed rates content in the form of a json
        """
        rates_content = rates_content.lstrip()
        rates_content = re.sub(",\s*", ",", rates_content)
        f = StringIO(rates_content)
        rates_df = pd.read_csv(
            f,
            names=["position", "sequence", "rate", "qq_interval", "std", "msa_data"],
            delimiter=r"\s+",
        )
        return rates_df

    @staticmethod
    def parse_output(output_path: str) -> t.Dict[str, t.Any]:
        """
        :return: None. parses the output file into a json form and saves it into self.result
        """
        with open(output_path, "r") as output_file:
            output_content = output_file.read()
        output_regex = re.compile(
            "#POS\s*SEQ\s*SCORE\s*QQ-INTERVAL\s*STD\s*MSA DATA.*?The alpha parameter (.\d*\.?\d*).*?LL=(-?\d*\.?\d*)(.*?)#Average",
            re.MULTILINE | re.DOTALL,
        )
        output_match = output_regex.search(output_content)
        result = dict()
        result["alpha"] = float(output_match.group(1))
        result["log_likelihood"] = float(output_match.group(2))
        result["rate_by_position"] = Rate4Site.parse_rates(output_match.group(3))
        return result
